onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlDFT.h

```diff
@@ -0,0 +1,702 @@
+            auto totalElementCount =
+                std::accumulate(constants.OutputSizes,
+                                constants.OutputSizes + _countof(constants.OutputSizes),
+                                1,
+                                std::multiplies<uint32_t>());
+            constants.ElementCount = totalElementCount / constants.OutputSizes[3];
+            constants.DFTIteration = index + 1;
+            Dispatch(in, out, constants, commandList);
+        }
+
+        // Transition resources to common state
+        barriers[0] = CD3DX12_RESOURCE_BARRIER::Transition(
+                inputResource,
+                D3D12_RESOURCE_STATE_UNORDERED_ACCESS,
+                D3D12_RESOURCE_STATE_COMMON
+                );
+
+        barriers[1] = CD3DX12_RESOURCE_BARRIER::Transition(
+                outputResource,
+                D3D12_RESOURCE_STATE_UNORDERED_ACCESS,
+                D3D12_RESOURCE_STATE_COMMON
+                );
+
+        commandList->ResourceBarrier(2, barriers);
+    }
+
+    void Dispatch(
+        ID3D12Resource* inputResource,
+        ID3D12Resource* outputResource,
+        DFTShaderConstants& constants,
+        ID3D12GraphicsCommandList* commandList)
+    {
+        D3D12_RESOURCE_BARRIER uav_barriers[2];
+        uav_barriers[0] = CD3DX12_RESOURCE_BARRIER::UAV(inputResource);
+        uav_barriers[1] = CD3DX12_RESOURCE_BARRIER::UAV(outputResource);
+        commandList->ResourceBarrier(2, uav_barriers);
+        // Set resource views
+        commandList->SetComputeRootUnorderedAccessView(
+            0, // root parameter index
+            inputResource->GetGPUVirtualAddress()
+        );
+
+        commandList->SetComputeRootUnorderedAccessView(
+            1, // root parameter index
+            outputResource->GetGPUVirtualAddress()
+        );
+        auto pendingElementCount = constants.ElementCount;
+
+        // Dispatch up to the maximum number of threads per iteration until
+        // all elements are completed
+        while (pendingElementCount > 0)
+        {
+            constants.StartIndex = constants.ElementCount - pendingElementCount;
+
+            uint32_t dispatchSizeX;
+
+            DFTHelpers::GetNextDispatchSize(
+                pendingElementCount,
+                1,
+                64,
+                dispatchSizeX,
+                pendingElementCount
+            );
+
+            // Set root constants
+            commandList->SetComputeRoot32BitConstants(
+                2, // root parameter index
+                21, // Constant count
+                &constants,
+                0 // offset
+            );
+
+            commandList->Dispatch(dispatchSizeX, 1, 1);
+        }
+
+        commandList->ResourceBarrier(2, uav_barriers);
+    }
+
+    void BluesteinZChirp(
+        ID3D12Resource* /*inputResource*/,
+        ID3D12Resource* /*outputResource*/,
+        ID3D12GraphicsCommandList* /*commandList*/)
+    {
+        // Not implemented
+        throw;
```

Comment:
>throw [](http://example.com/codeflow?start=8&length=5)

🤨 What are we throwing here?

We should use the `ORT_THROW_X` macros rather than direct throw calls, since some clients want throws to resolve to `std::abort` when `ORT_NO_EXCEPTIONS` is defined.